warehouse/query-core/src/main/java/datawave/query/jexl/nodes/QueryPropertyMarker.java

```diff
@@ -161,6 +165,66 @@ protected void setupSource(JexlNode source) {
         this.jjtAddChild(refExpNode1, 0);
     }
     
+    /**
+     * Determines if this source node is already marked with the provided marker class
+     *
+     * @param source
+     *            a JexlNode
+     * @param markerClass
+     *            an instance of a {@link QueryPropertyMarker}
+     * @return true if this node is already marked with the provided marker class
+     */
+    public static boolean isSourceAlreadyMarked(JexlNode source, Class<? extends QueryPropertyMarker> markerClass) {
+        return QueryPropertyMarker.findInstance(source).isType(markerClass);
+    }
+    
+    /**
+     * Determine if one the source node's ancestors is already marked. For example, is this subtree already marked as delayed?
+     *
+     *
+     * @param node
+     *            a JexlNode
+     * @return true
+     */
+    public static boolean isSubTreeAlreadyMarked(JexlNode node, Class<? extends QueryPropertyMarker> markerClass) {
```

Comment:
The original source node is passed into this, which is then subsequently checked to see if it's marked before checking for a marked ancestor. Do we really need isSourceMarked() given this?

Additionally, I don't love `isSubTreeAlreadyMarked()` as a name, to my mind that implies that we're descending into the source's children rather than ascending through the source's parentage. How do you feel about `isPartOfMarkedSubTree()` or something similar? Alternatively, we could pass in the source's parent, check for null first, and rename this to `hasMarkedAncestor()`.

Additionally pt. 2, this can be a private method.

Reply:
I like renaming the method to `hasMarkedAncestor` or `isAncestorMarked` and making the related changes you suggested.

Ideally, both `hasMarkedAncestor` and `isSourceMarked` would be methods available to any caller. I can imagine a situation where a visitor would need to know if an ancestor is marked or not before modifying the query tree.

Reply:
You make a good point about leaving it public for outside callers to use. I would add a note to the documentation for `isSubTreeAlreadyMarked` to stress that it will traverse up the node's entire ancestry, and not just the immediate ancestor to make a note of comparative expense.

Reply:
I'll update the docs